Reporting the state of free item positioning mode

Free item positioning mode drastically changes how Reaper behaves. Ergo, it would be beneficial to have this reported in two ways:

1. Feedback for the following toggle:
Action: Track properties: Toggle free item positioning
ID: 40641

2. Can we be notified if FIPM is on when moving to a track?

As ever, thanks in advance.

In a fit of sleep-deprived ambition, I've attempted to implement action feedback for the FIPM toggle action above, but when I build, scons is disappointed in me:
src\reaper_osara.cpp(1827): error C2065: 'postToggleFreeItemPositioning': undeclared identifier
Does anyone have time to nudge a newb toward declaring this? My intentions are good, I'll add it to the ReadMe and everything. :)

--- a/readme.md
+++ b/readme.md
@@ -6,7 +6,7 @@
 
 OSARA is a [REAPER](http://www.reaper.fm/) extension which aims to make REAPER accessible to screen reader users.
 It was heavily inspired by and based on the concepts of the ReaAccess extension, but was created as a potential replacement because ReaAccess seems to have been abandoned and was not developed openly.
-It runs on both Windows and Mac, though it is currently very experimental on Mac ad is not yet fully functional.
+It runs on both Windows and Mac, though it is currently very experimental on Mac and is not yet fully functional.
 
 Features:
 
@@ -89,6 +89,7 @@ Most of these are actions built into REAPER, but a few are very useful actions f
 - SWS: Select nearest previous folder: Alt+PageUp
 - Track: Remove tracks
 - Track: Select all tracks
+- Track properties: Toggle free item positioning
 
 #### Adjusting Track Parameters
 - Track: Toggle mute for selected tracks

--- a/src/reaper_osara.cpp
+++ b/src/reaper_osara.cpp
@@ -3,6 +3,7 @@
  * Main plug-in code
  * Author: James Teh <jamie@jantrid.net>
  * Copyright 2014-2019 NV Access Limited, James Teh
+ * Copyright 2019-2021 James Teh
  * License: GNU General Public License version 2.0
  */
 
@@ -441,6 +442,10 @@ bool isItemSelected(MediaItem* item) {
 	return *(bool*)GetSetMediaItemInfo(item, "B_UISEL", NULL);
 }
 
+bool isFreeItemPositioningEnabled(MediaTrack* track) {
+	return *(int*)GetSetMediaTrackInfo(track, "B_FREEMODE", NULL);
+}
+
 const char* automationModeAsString(int mode) {
 	// this works for track automation mode and global automation override.
 	switch (mode) {
@@ -694,6 +699,9 @@ void postGoToTrack(int command, MediaTrack* track) {
 		// "2 items".
 		s << " " << format(translate_plural("{} item", "{} items", itemCount),
 			itemCount);
+	}
+	if (isFreeItemPositioningEnabled(track)) {
+		s << " " << translate("free item positioning");
 	}
 	if (isTrackGrouped(track)) {
 		// Translators: Reported when navigating to a track which is grouped.
@@ -1661,6 +1669,16 @@ void postToggleEnvelopePointsMoveWithMediaItems(int command) {
 		translate("disabled envelope points move with media items"));
 }
 
+void postToggleFreeItemPositioning(int command) {
+	MediaTrack* track = GetLastTouchedTrack();
+	if (!track) {
+		return;
+	}
+	outputMessage(*(bool*)GetSetMediaTrackInfo(track, "B_FREEMODE", nullptr) ?
+		translate("enabled free item positioning") :
+		translate("disabled free item positioning"));
+}
+
 typedef void (*PostCommandExecute)(int);
 typedef struct PostCommand {
 	int cmd;
@@ -1823,6 +1841,7 @@ PostCommand POST_COMMANDS[] = {
 	{40218, postChangeTransientDetectionThreshold}, // Transient detection threshold: Increase
 	{40219, postChangeTransientDetectionThreshold}, // Transient detection threshold: Decrease
 	{40070, postToggleEnvelopePointsMoveWithMediaItems}, // Options: Envelope points move with media items
+	{40641, postToggleFreeItemPositioning}, // Track properties: Toggle free item positioning
 	{0},
 };
 MidiPostCommand MIDI_POST_COMMANDS[] = {

--- a/src/reaper_osara.cpp
+++ b/src/reaper_osara.cpp
@@ -2,8 +2,7 @@
  * OSARA: Open Source Accessibility for the REAPER Application
  * Main plug-in code
  * Author: James Teh <jamie@jantrid.net>
- * Copyright 2014-2019 NV Access Limited, James Teh
- * Copyright 2019-2021 James Teh
+ * Copyright 2014-2021 NV Access Limited, James Teh
  * License: GNU General Public License version 2.0
  */
 

